### PR TITLE
Improve description of how =trace is used

### PR DESCRIPTION
**Pull request by GordonBGood**
The following nuances weren't previously fully explained:

1. That `=trace` is only used by `--gc:orc`.
2. That `=trace` is almost certainly used along with `=destroy` when manual resource allocation has been used, but it is only required if there is a possibility of cyclic references in the wrapped types (ie. generic types).
3. That, currently, a forward definition is required for the second of the pair to avoid an auto compiler generation conflict.

The pattern of the use of `=trace` has also been made more extensive, showing how both a custom `=destroy` and `=trace` are used for manual allocation of resources when there is any possibility of a cyclic reference in the resource-wrapped values.

### CODE REVIEW

File: doc/destructors.rst [hunk 1]
```diff
@@ -222,16 +222,34 @@ The prototype of this hook for a type `T` needs to be:
...
-      `=trace`(child, env)
+  type
+    Test[T] = object
+      size: Natural
+      arr: ptr UncheckedArray[T] # raw pointer field
+
+  proc makeTest[T](size: Natural): Test[T] = # custom allocation...
+    Test[T](size: size, arr: cast[ptr UncheckedArray[T]](alloc0(sizeof(T) * size)))
+
+  proc `=trace`[T](dest: var Test[T]; env: pointer) # forward declaration
```
Comment (Araq) [inline doc/destructors.rst]: Why is this required?

Comment (GordonBGood) [reply doc/destructors.rst]: @Araq:

> Why is this required?

While I was doing  "Ideally you describe the issue in destructors.rst", I thought I would just give a more complete example of how one would need to use **both** `=destroy` and `=trace` when manual resource allocation was done in the case that there might be cyclic references in the contained manually allocated data.

So the `makeTest` proc illustrates that there is a manual memory allocation to an internal pointer field, and the forward declaration of `=trace` is there because currently it won't work with both `=destroy` and `=trace` hooks without forward declaring the second of them, which is the issue.

Try it, I think you will find it won't work without the forward declaration.  If you have another way, then cancel the PR and let's document the preferred way.

Comment (Araq) [reply doc/destructors.rst]: Well I have a working test case with `=trace` and all it takes is to declare =destroy first.

Comment (GordonBGood) [reply doc/destructors.rst]: @Araq:

I tried that and got the opposite problem of `=trace` complaining that it had already been created inside `=destroy` when there truly was cyclic data such as the Hammings Test.  I added the completely unnecessary `=trace` to the Hamming Numbers Test as follows (unnecessary because the compiler generated `=trace` already handles the destruction since it uses nothing but regular `ref`'s; in fact the custom `=destroy` is also completely unnecessary for the same reason and is just there to help show that as many objects are being created as destroyed):

```nim
proc `=trace`(ll: var LazyListObj; env: pointer) # forward declaration

proc `=destroy`(ll: var LazyListObj) =
  when defined(trace20):
    echo "destroying:  ", (destroyed, ll.hd[1].convertTriVal2BigInt, ll.tlf == nil, ll.tl == nil)
  destroyed += 1
  if ll.tlf != nil: ll.tlf.`=destroy`
  if ll.tl != nil: ll.tl.`=destroy`

proc `=trace`(ll: var LazyListObj; env: pointer) =
  when defined(trace20):
    echo "tracing:  ", (destroyed, ll.hd[1].convertTriVal2BigInt, ll.tlf == nil, ll.tl == nil)
  if ll.tlf != nil: ll.tlf.`=trace` env
  if ll.tl != nil: ll.tl.`=trace` env
```
which works as before; I originally wrote this to better understand `=trace` but actually it serves to function as a test showing that custom `=trace` proc's work.

However, if I comment out the forward definition, I get the following: "Error: cannot bind another '=trace' to: LazyListObj; previous declaration was constructed here implicitly", where "here" refers to the line in `=destroy` that contains the possibility for the reference back through to the same `LazyListObj` at the tail.

In other words, you don't need it for the simplified example as the compiler can't realize that one might put `T` as a reference to itself, but for things like lists where the type of the internal tail field is exactly the same type that is both being destroyed and traced, then there is this potential problem.

I didn't want to fill the text with this long explanation, but just showed the use example of how it will always be able to be used no matter the chain of reference.

Comment (GordonBGood) [reply doc/destructors.rst]: @Araq:
I've been trying to come up with a simpler example.  Some examples aren't usable anyway because they result in the halting problem if there isn't the laziness injected as per the LazyList example.  This one case of the Hamming Numbers Test is a bit artificial in that the use of the custom `=trace` is something I came up with for testing and understanding and isn't something that would normally come up in real use.

So I guess if the need for the forward definition is so rare, then it maybe doesn't need to be documented or mentioned?  The forward definition doesn't hurt anything, so the question is whether to leave it in or take it out?

Comment (Araq) [reply doc/destructors.rst]: > So I guess if the need for the forward definition is so rare, then it maybe doesn't need to be documented or mentioned? 

That's what I think too, yes.

Comment (GordonBGood) [reply doc/destructors.rst]: > > So I guess if the need for the forward definition is so rare, then it maybe doesn't need to be documented or mentioned?

> That's what I think too, yes.

It really is rare: the Hamming Numbers Test example (or something similar) is the only way I've been able to reproduce the need for the forward declaration.

**EDIT_ADD:**
I am thinking that it possibly is triggered when closures containing references to cyclic data are used as for the "Thunk" operations in the implementation of the LazyList in the Hamming test may be the cause, but I can't be bothered to trace it down further since it is so rare and as the work around works well enough.

A minimum more complete example that doesn't need the forward declaration but does require the custom "hooks" due to the use of the pointer is the following code, modelled after the issue #18421 that started all of this:

**EDIT_ADD:**  more complex pointer contents in keeping with the original example.

```nim
type
  CycleObj[T] = object
    size: Natural
    contentp: ptr UncheckedArray[T]
    child: Cycle[T]    
  Cycle[T] = ref CycleObj[T]

proc `=destroy`[T](co: var CycleObj[T]) =
  if co.contentp != nil:
    for i in 0 ..< co.size: co.contentp[i].`=destroy`
    co.contentp.deallocShared; co.contentp = nil
  co.child.`=destroy`

proc `=trace`[T](co: var CycleObj[T], env: pointer) =
  if co.contentp != nil:
    for i in 0 ..< co.size: co.contentp[i].`=trace` env
  co.child.`=trace` env

proc newCycle[T](oa: openArray[T]): Cycle[T] =
  let sz = oa.len
  result =
    Cycle[T](contentp: cast[ptr UncheckedArray[T]](allocShared0(sizeof(T)*sz)))
  result.size = sz
  for i in 0 ..< sz: result.contentp[i] = oa[i]

proc main =
  var c = newCycle([ 4, 13, 42 ])
  c.child = c
  echo c.contentp[2]
  c = nil; GC_FullCollect() # somewhat deterministic trace and destroy

when isMainModule:
  let mem = getOccupiedMem()
  main()
  GC_FullCollect() # manual trigger; auto triggered on memory threshold
  echo "Memory balance:  ", getOccupiedMem() - mem
```
So how should we proceed?

1. Just eliminate the forward declaration from this PR?
2. Change the PR to give a more complete example such as the above?
3. Accept the PR as it is as the forward declaration doesn't do any harm?

Perhaps the PR should be augmented to mention that the trace and destroy operations for cyclic references only occur when `GC_FullCollect()` is called or automatically triggered when a memory threshold is reached?  I was unclear on this until I did the experiments.  The comments in the above code mention this extra information.

Comment (Araq) [reply doc/destructors.rst]: > Just eliminate the forward declaration from this PR?

This one.

Comment (Araq) [reply doc/destructors.rst]:
```suggestion
```
